resilience4j-circuitbreaker/src/main/java/io/github/resilience4j/circuitbreaker/CircuitBreaker.java

```diff
@@ -997,4 +1014,112 @@ static CircuitBreaker of(String name, Supplier<CircuitBreakerConfig> circuitBrea
+     *
+     * @param <T> of return type
+     */
+    final class CircuitBreakerFuture<T> implements Future<T> {
+        final private Future<T> future;
+        final private OnceConsumer<CircuitBreaker> onceToCircuitbreaker;
+        final private NanosecondTimer nanosecondTimer;
+
+        CircuitBreakerFuture(CircuitBreaker circuitBreaker, Future<T> future) {
+            this(circuitBreaker, future, now());
+        }
+
+        CircuitBreakerFuture(CircuitBreaker circuitBreaker, Future<T> future, long start) {
+            this(circuitBreaker, future, create(start));
+        }
+
+        CircuitBreakerFuture(CircuitBreaker circuitBreaker, Future<T> future, NanosecondTimer timer) {
+            Objects.requireNonNull(future, "Non null Future is required to decorate");
+            this.onceToCircuitbreaker = OnceConsumer.of(circuitBreaker);
+            this.future = future;
+            this.nanosecondTimer = timer;
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return future.cancel(mayInterruptIfRunning);
+        }
+
+        @Override
+        public boolean isCancelled() {
+            return future.isCancelled();
+        }
+
+        @Override
+        public boolean isDone() {
+            return future.isDone();
+        }
+
+        @Override
+        public T get() throws InterruptedException, ExecutionException {
+            try {
+                T v = future.get();
+                onceToCircuitbreaker.applyOnce(cb -> nanosecondTimer.elapsed(
+                        time -> cb.onSuccess(time, TimeUnit.NANOSECONDS)));
+                return v;
+            } catch (CancellationException | InterruptedException e) {
+                onceToCircuitbreaker.applyOnce(cb -> cb.releasePermission());
+                throw e;
+            } catch (Exception e) {
+                onceToCircuitbreaker.applyOnce(cb -> nanosecondTimer.elapsed(
+                        time -> cb.onError(time, TimeUnit.NANOSECONDS, e)));
+                throw e;
+            } finally {
+                future.cancel(true);
+            }
+        }
+
+        @Override
+        public T get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+            try {
+                T v = future.get(timeout, unit);
+                onceToCircuitbreaker.applyOnce(cb -> nanosecondTimer.elapsed(
+                        time -> cb.onSuccess(time, TimeUnit.NANOSECONDS)));
+                return v;
+            } catch (CancellationException | InterruptedException e) {
+                onceToCircuitbreaker.applyOnce(cb -> cb.releasePermission());
+                throw e;
+            } catch (Exception e) {
+                onceToCircuitbreaker.applyOnce(cb -> nanosecondTimer.elapsed(
+                        time -> cb.onError(time, TimeUnit.NANOSECONDS, e)));
+                throw e;
+            } finally {
+                future.cancel(true);
```

Comment:
In get(timeout) when timeout occurs, caller decide to walks away and requesting task to stop before leaving. (its more of good practice in evaluating future)

Reply:
Can we catch the `TimeoutException`. I think it's missing. And can we just cancel then?

Reply:
Sure, I will make that change.

Reply:
We think that we should remove `future.cancel(true);` altogether. Reason: if multiple clients share the same `Future` instance and one says `get(5, SECONDS)`, other `get(30, SECONDS)` and computation completes in 10 seconds than the later one would not get results because of the cancellation done by the first one.